ClassicAssertExtensions reestablish the classic Assert as nunit.framework.Assert class members.

OsirisTerje opened the pull request:
Fixes #4693 
Fixes #4613

The introduction of the new extension mechanism in C14 allow extension methods on static classes.  Using this we can create extension methods on the Assert and Assume static classes that "moves" the classic asserts back.  The extension methods live in the legacy framework project, so they don't interfere with the nunit.framework project itself.  

This is what we really wanted when we had to separate the class asserts from the constraint model in NUnit V4.

1. This works for projects with C14/SDK10.  Trying to use these with <C14 give a compiler error clearly stating you need C14.
2. NUnit projects require C14/SDK10
3. Added verification tests

## Implementation

The extension mechanism for the Assert class is done like this  (prototype):

<img width="1948" height="787" alt="image" src="https://github.com/user-attachments/assets/67b39024-bc38-4204-9862-d33e67848478" />

When it is run with a failing test, you can see  (1) it looks like it belong to the framework Assert class and 2) we see it as an constraint test - because it still IS a classic assert (which as known, calls the constraint model). 

<img width="2117" height="1033" alt="image" src="https://github.com/user-attachments/assets/f544203b-96d2-4d7a-b449-d8d3c623cf2d" />

And as (1) above, you can see it is a classic method:

<img width="1546" height="558" alt="image" src="https://github.com/user-attachments/assets/a385d3ed-012e-4297-aea2-37f9a5d3b9f0" />

manfred-brands commented on src/NUnitFramework/nunit.framework.legacy/LegacyExtensions.Comparisons.cs:
```diff
@@ -0,0 +1,144 @@
+        {
+            #region Greater
+
+            public static void Greater(int arg1, int arg2, string? message = null, params object[] args) => Legacy.ClassicAssert.Greater(arg1, arg2, message ?? string.Empty, args);
```
3.14 doesn't have nullable _message_. It has two overloads, one with and one without _message_ parameter.
Should we not do the same here?

manfred-brands commented on src/NUnitFramework/nunit.framework.legacy/LegacyExtensions.Numeric.cs:
```diff
@@ -0,0 +1,130 @@
+            #region Zero
+
+            public static void Zero(int actual) => Legacy.ClassicAssert.Zero(actual);
+            public static void Zero(int actual, string? message, params object?[]? args) => Legacy.ClassicAssert.Zero(actual, message ?? string.Empty, args);
```
Interesting, 3.14 does have a nullable _message_ here.

manfred-brands commented on src/NUnitFramework/nunit.framework.legacy/LegacyExtensions.Types.cs:
```diff
@@ -0,0 +1,80 @@
+                => Legacy.ClassicAssert.IsAssignableFrom(expected, actual, message ?? string.Empty, args);
+
+            public static void IsAssignableFrom(Type expected, object? actual)
+                => Legacy.ClassicAssert.IsAssignableFrom(expected, actual);
```
The _message_ should not be optional, you already have the overload without a _message_ parameter.

manfred-brands commented on src/NUnitFramework/nunit.framework.legacy/LegacyExtensions.cs:
```diff
@@ -0,0 +1,356 @@
+            /// <summary>
+            /// Verifies that two doubles are equal considering a delta.
+            /// </summary>
+            public static void AreEqual(double expected, double actual, double delta, string? message = null, params object?[]? args)
```
_message_ should not be optional.

OsirisTerje commented on src/NUnitFramework/nunit.framework.legacy/nunit.framework.legacy.csproj:
```diff
@@ -3,7 +3,8 @@
     <RootNamespace>NUnit.Framework.Legacy</RootNamespace>
-    <NoWarn/>
+    <LangVersion>latest</LangVersion>
+    <NoWarn>CS1591;IDE0005</NoWarn>
```
Agree

stevenaw commented on src/NUnitFramework/Directory.Build.props:
```diff
@@ -1,7 +1,7 @@
 
   <PropertyGroup>
-    <LangVersion Condition="'$(MSBuildProjectExtension)' == '.csproj'">13</LangVersion>
+    <LangVersion Condition="'$(MSBuildProjectExtension)' == '.csproj'">14</LangVersion>
```
Good catch. It looks like we missed this earlier.
Should we continue to target a specific version here, or is `latest` sufficient?

stevenaw commented:
Thanks @OsirisTerje 
I was able to get through a few files before having to step away. I left a few questions

OsirisTerje commented on src/NUnitFramework/nunit.framework.legacy/LegacyExtensions.Collection.cs:
```diff
@@ -0,0 +1,133 @@
+using System.Collections;
+
+// ReSharper disable once CheckNamespace
+namespace NUnit.Framework
```
It will be in the NUnit.Framework namespace, and connected to the Assert there, but the implementation is in the Framework.Legacy project.   So there is no opt-in, it will just be there, and pop up on Intellisense when you are on C# 14.   Pretty cool actually :-) 
Now we distribute the legacy together with the framework, so all should be good. 


This is what we hoped should be ready earlier for V4, but which was delayed by MS. Extension methods on static classes.  Now it finally is in. 

And we should of course mention and thank [Joseph](https://github.com/jnm2) - this is actually his invention, as he worked with the C# language group.

stevenaw replied:
Ah yes, I remember this coming up now as part of our discussions for how to handle the original split. Thanks!

OsirisTerje commented:
@manfred-brands I believe there should be 1:1 now.  Also renamed from Legacy to Classic extensions. Sounds better.  Inheritdoc fixed, and warning suppressions removed.
Added also a document there  classic_extensions.md to verify we got them all.

manfred-brands commented on src/NUnitFramework/nunit.framework.legacy.tests/AssertWrapperSmokeTests.cs:
I think we can delete this file as you now have explicit tests.

manfred-brands commented on src/NUnitFramework/nunit.framework.legacy/nunit.framework.legacy.csproj:
```diff
@@ -3,7 +3,6 @@
   <PropertyGroup>
     <TargetFrameworks>$(NUnitLibraryFrameworks)</TargetFrameworks>
     <RootNamespace>NUnit.Framework.Legacy</RootNamespace>
-    <NoWarn/>
```
Reset this as now it inherits from Directory.Build.props which as a NoWarn of 1591 we use for our test projects.

manfred-brands replied:
Done.

manfred-brands reviewed (approved):
@OsirisTerje Copilot got overzealous, it created Assert extensions for method that were never part of Assert such as CollectionAssert, StringAssert etc.

I have removed those and updated the .md document.
I have also addressed the formatting issues.

I compared the Assert with the extensions and both have 184 methods.
